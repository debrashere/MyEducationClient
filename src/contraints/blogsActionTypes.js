export const CREATE_BLOG = 'CREATE_BLOG';     
export const CREATE_BLOG_SUCCESS = 'CREATE_BLOG_SUCCESS';
export const UPDATE_BLOG=  'UPDATE_BLOG';
export const UPDATE_BLOG_SUCCESS = 'UPDATE_BLOG_SUCCESS';
export const FETCH_BLOGS= 'FETCH_BLOGS';   
export const FETCH_BLOGS_SUCCESS = 'FETCH_BLOGS_SUCCESS';
export const FETCH_BLOG = 'FETCH_BLOG';
export const FETCH_BLOG_SUCCESS = 'FETCH_BLOG_SUCCESS';
export const DELETE_BLOG = 'DELETE_BLOG';
export const DELETE_BLOG_SUCCESS = 'DELETE_BLOG_SUCCESS';
export const BLOG_ERROR= 'BLOG_ERROR'; 

