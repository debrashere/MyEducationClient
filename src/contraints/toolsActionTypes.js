export const CREATE_TOOL = 'CREATE_TOOL';
export const CREATE_TOOL_SUCCESS = 'CREATE_TOOL_SUCCESS';
export const UPDATE_TOOL = 'UPDATE_TOOL';
export const UPDATE_TOOL_SUCCESS = 'UPDATE_TOOL_SUCCESS';
export const FETCH_TOOL = 'FETCH_TOOL';
export const FETCH_TOOL_SUCCESS = 'FETCH_TOOL_SUCCESS';
export const FETCH_TOOLS = 'FETCH_TOOLS';
export const FETCH_TOOLS_SUCCESS = 'FETCH_TOOLS_SUCCESS';
export const TOOLS_ERROR = 'TOOLS_ERROR';
export const SET_TOOL = 'SET_TOOL';
export const SET_TOOL_SUCCESS = 'SET_TOOL_SUCCESS';