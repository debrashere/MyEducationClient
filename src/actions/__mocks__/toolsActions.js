import * as types from '../../contraints/toolsActionTypes';
//import * as actions from '../../actions/toolsActions';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

let tools = [];
let mockedStore = mockStore({ auth: {authToken:'token' }});
 
export const setTools = (thisTools) => ( tools = thisTools );
export const setStore = (thisStore) => ( mockedStore = thisStore );
export const store = mockedStore;

export const fetchTools = () => (dispatch, getState) => {    

    global.fetch = jest.fn().mockImplementation(() =>
                Promise.resolve({
                    ok: true,
                    json() {
                        return tools;
                    }
                })
            );

            const expected = [
                {type: types.FETCH_TOOLS},
                {tools: tools,
                type: types.FETCH_TOOLS_SUCCESS}
            ]
            
            return expected        
}
    
 
