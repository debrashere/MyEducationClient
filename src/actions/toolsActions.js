import {SubmissionError} from 'redux-form';
import {API_BASE_URL} from '../config';
import {normalizeResponseErrors} from './utils';
import * as types from '../contraints/toolsActionTypes';

export const toolsError = error => ({
    type: types.TOOLS_ERROR,
    error
});

export const fetchToolsSuccess = (tools) => ({
    type: types.FETCH_TOOLS_SUCCESS,
    tools
});

export const updateToolSuccess = (tool) => ({
    type: types.UPDATE_TOOL_SUCCESS,
    tool
});

export const fetchToolSuccess = (tool) => {
return ({  
    type: types.FETCH_TOOL_SUCCESS,
    tool: tool
});
}

export const createToolSuccess = (tool) => ({
    type: types.CREATE_TOOL_SUCCESS,
    tool
});

export const createTool = (userName, title, url, description, price, rating) =>  (dispatch, getState) => {
    const authToken = getState().auth.authToken;
    dispatch({type: types.CREATE_TOOL});
    fetch(`${API_BASE_URL}/tools`, {
       method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            // Provide our auth token as credentials
            Authorization: `Bearer ${authToken}`
        },
        body: JSON.stringify({ title, url, description, price, rating})
    })      
       .then(response => response.json())      
       .then(tool => { 
           dispatch(createToolSuccess(tool));
       })
       .catch(err => {
           const {code} = err;
           const message =
               code === 401
                   ? 'Unable to create tool at this time'
                   : 'Failed to create tool, try again';
           dispatch(toolsError(message));
           throw new SubmissionError({             
            _error:  message
          }) 
       })
};

export const updateTool = ( id, title, url, description, price, rating) =>  (dispatch, getState) => {
    const authToken = getState().auth.authToken; 
    dispatch({type: types.UPDATE_TOOL});   
    return fetch(`${API_BASE_URL}/tools/${id}`, {
        method: 'PUT',
        headers: {
            'Content-Type': 'application/json',
            // Provide our auth token as credentials
            Authorization: `Bearer ${authToken}`
        },
        body: JSON.stringify({ id, title, url, description, price, rating})
    })
        .then(res => normalizeResponseErrors(res))
        .then(res => res.json())
        .then(tool => {    
            dispatch(updateToolSuccess(tool)) })        
        .catch(err => {
            const {reason, message} = err;
            dispatch(toolsError(message));
            if (reason === 'ValidationError') {
                throw new SubmissionError({             
                    _error:  message
                  }) 
            }
        });
};

export const fetchTools = () => (dispatch, getState) => {
    dispatch({type:types.FETCH_TOOLS});
    const authToken = getState().auth.authToken;
 
    return  (
         fetch(`${API_BASE_URL}/tools`, {
            method: 'GET',
            headers: {
                // Provide our auth token as credentials
                Authorization: `Bearer ${authToken}`
            }
        })
        .then(response => response.json())
        .then(tools => { 
            dispatch(fetchToolsSuccess(tools)) })
        .catch(err => {
            const {code} = err;
            const message =
                code === 401
                    ? 'Your are not authorize.  Please log in.'
                    : 'Failed to fetch posts tool, try again';
            dispatch(toolsError(message));
            throw new SubmissionError({             
                _error:  message
              })         
        })
    )
 };

 export const fetchTool = (id) => (dispatch, getState) => {
     dispatch({type:types.FETCH_TOOL});
     const authToken = getState().auth.authToken;
     return  (
          fetch(`${API_BASE_URL}/tools/${id}`, {
             method: 'GET',
             headers: {
                 // Provide our auth token as credentials
                 Authorization: `Bearer ${authToken}`
             }           
         })
         .then(res => normalizeResponseErrors(res))
         .then(res => res.json())
         .then(tool => {                
             dispatch(fetchToolSuccess(tool));
             return Promise.resolve();                                 
         })         
         .catch(err => {
             const {code} = err;
             const message =
                 code === 401
                     ? 'Your are not authorize.  Please log in.'
                     : 'Failed to fetch posts tool, try again';
             dispatch(toolsError(message));
             return Promise.reject(
                 new SubmissionError({
                     _error: message
                 })
             );
         })
     )
  };
