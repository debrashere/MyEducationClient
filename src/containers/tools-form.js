import React from 'react';
import {connect} from 'react-redux';
import {Field, reduxForm, focus } from 'redux-form';
import {Link} from 'react-router-dom';
import {createTool} from '../actions/toolsActions';
import Input from '../components/input';
import {required, nonEmpty, isTrimmed, isUrlFormatValid, isNumeric, isRatingValid} from '../validators';

export class ToolsForm extends React.Component {  
    // Submit the new tool data  
 
    onSubmit(values) {
        return this.props
            .dispatch(createTool(this.props.currentUser.username, values.title,
                values.url, values.description, values.price,
                values.rating))                               
    }

    render() {
        const { error, handleSubmit, pristine, reset, submitting, submitSucceeded } = this.props
        let formattedError;     
        if (error) {
            formattedError =             
            <div className="message-error">
                {<strong>{error}</strong>}                 
            </div>
        }
        let successMessage;
        if (submitSucceeded === true){
            successMessage =  
            <div className="message-success">
              <strong>Submission successful</strong>}                 
            </div>
        }
        return ( 
            <div> 
                <h2>Tools Form</h2>   
                <form className="add-form form" 
                    onSubmit={handleSubmit(values => this.onSubmit(values) )}> 

                    {formattedError} 
                    {successMessage}   
                    <label htmlFor="title">Title</label>
                    <Field
                        component={Input}
                        type="text"
                        name="title"
                        validate={[required, nonEmpty, isTrimmed]}
                    />
                    
                    <label htmlFor="description">Description</label>
                    <Field component={Input}
                        type="text" 
                        name="description"
                        validate={[required, nonEmpty, isTrimmed]}
                    />

                    <label htmlFor="url">Url</label>
                    <Field
                        component={Input}
                        type="text"
                        name="url"
                        validate={[required, isTrimmed, isUrlFormatValid]}
                    />

                    <label htmlFor="price">Price</label>
                        <Field component={Input}
                        type="text" 
                        name="price"
                        validate={[required, nonEmpty, isTrimmed, isNumeric]}
                    />

                    <label htmlFor="rating">Rating</label>
                        <Field component={Input} type="text" name="rating"
                        validate={[required, isTrimmed, isNumeric, isRatingValid]}
                    />
                                
                    <div>
                        <button
                            className="add-button"
                            type="submit"
                            disabled={pristine || submitting}>
                            Add Tool
                        </button>
                        <button 
                            className="clear-button"
                            type="button" 
                            disabled={pristine || submitting} onClick={reset}>
                            Clear Values
                        </button>&nbsp;&nbsp;
                        <Link to="/tools">(Back to tools page)</Link>
                    </div>
        
                </form>
            </div>             
        )
    }
}

export const mapStateToProps = state => ({
    currentUser:{
        firstName: "beebee",
        lastName: "sanders",
        username: "bsanders" }
        // state.auth.currentUser
});

ToolsForm = connect(
    mapStateToProps
  )(ToolsForm)

export default reduxForm({
    form: 'toolsForm',    
    onSubmitFail: (errors, dispatch) => dispatch(focus('toolsform', 'title'))        
})(ToolsForm);
