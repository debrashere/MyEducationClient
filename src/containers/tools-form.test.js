import React from 'react';
import {shallow, mount} from 'enzyme';
import * as types from '../contraints/toolsActionTypes';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter as Router, Switch, Route, Link } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import {mapStateToProps, ToolsForm, reduxForm} from './tools-form';
import renderer  from 'react-test-renderer';
 
jest.mock('../actions/toolsActions');
import * as actions from '../actions/__mocks__/toolsActions';
import { setStore, store} from  '../actions/__mocks__/toolsActions';
 
//configure({ adapter: new Adapter() });
const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)
   
describe('<ToolsForm />', () => {
    let thisTools = [];
    let currentUser;
    let loggedIn;
    let props;
    let values;
    beforeAll(() => {

        for (let i = 1; i < 6 ; i++) {
            thisTools.push({
                id: i,
                title: `title-${i}`,
                description: `description-${i}`,
                url: `url${i}`,
                price: i ,
                rating: i,
            });
        }

        currentUser = {
            firstName: "beebee",
            lastName: "sanders",
            username: "bsanders" };

        loggedIn  = false;

        setStore(mockStore({ tools: [],  auth: {authToken:'token' }}));
        
        values = { 
            title: 'title',
             url: 'www.google.com', 
             description: 'description', 
             price: '1.00',
             rating: '3' };

        let handleSubmit = fn => fn(values);
      
        props = {
                currentUser: currentUser,
                tools: thisTools,
                loggedIn: loggedIn,
                store,
                error: null,
                handleSubmit,
                pristine: true, 
                reset: false, 
                submitting: false, 
                submitSucceeded :null
            };
    })

    it('Renders without crashing', () => {
        shallow(<ToolsForm store={store} {...props} />);
    });
 
    it('Snapshot test', () => {
        const tree = renderer.create(<reduxForm {...props} />);   
        expect(tree.toJSON()).toMatchSnapshot();
    });
 
});
