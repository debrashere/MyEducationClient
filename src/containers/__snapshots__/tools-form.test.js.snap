// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToolsForm /> Snapshot test 1`] = `
<reduxForm
  currentUser={
    Object {
      "firstName": "beebee",
      "lastName": "sanders",
      "username": "bsanders",
    }
  }
  error={null}
  handleSubmit={[Function]}
  loggedIn={false}
  pristine={true}
  reset={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  submitSucceeded={null}
  submitting={false}
  tools={
    Array [
      Object {
        "description": "description-1",
        "id": 1,
        "price": 1,
        "rating": 1,
        "title": "title-1",
        "url": "url1",
      },
      Object {
        "description": "description-2",
        "id": 2,
        "price": 2,
        "rating": 2,
        "title": "title-2",
        "url": "url2",
      },
      Object {
        "description": "description-3",
        "id": 3,
        "price": 3,
        "rating": 3,
        "title": "title-3",
        "url": "url3",
      },
      Object {
        "description": "description-4",
        "id": 4,
        "price": 4,
        "rating": 4,
        "title": "title-4",
        "url": "url4",
      },
      Object {
        "description": "description-5",
        "id": 5,
        "price": 5,
        "rating": 5,
        "title": "title-5",
        "url": "url5",
      },
    ]
  }
/>
`;

exports[`<ToolsForm /> Snapshot test II 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <reduxForm
    currentUser={
        Object {
            "firstName": "beebee",
            "lastName": "sanders",
            "username": "bsanders",
          }
    }
    error={null}
    handleSubmit={[Function]}
    loggedIn={false}
    pristine={true}
    reset={false}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
    submitSucceeded={null}
    submitting={false}
    tools={
        Array [
            Object {
              "description": "description-1",
              "id": 1,
              "price": 1,
              "rating": 1,
              "title": "title-1",
              "url": "url1",
            },
            Object {
              "description": "description-2",
              "id": 2,
              "price": 2,
              "rating": 2,
              "title": "title-2",
              "url": "url2",
            },
            Object {
              "description": "description-3",
              "id": 3,
              "price": 3,
              "rating": 3,
              "title": "title-3",
              "url": "url3",
            },
            Object {
              "description": "description-4",
              "id": 4,
              "price": 4,
              "rating": 4,
              "title": "title-4",
              "url": "url4",
            },
            Object {
              "description": "description-5",
              "id": 5,
              "price": 5,
              "rating": 5,
              "title": "title-5",
              "url": "url5",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "currentUser": Object {
        "firstName": "beebee",
        "lastName": "sanders",
        "username": "bsanders",
      },
      "error": null,
      "handleSubmit": [Function],
      "loggedIn": false,
      "pristine": true,
      "reset": false,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "submitSucceeded": null,
      "submitting": false,
      "tools": Array [
        Object {
          "description": "description-1",
          "id": 1,
          "price": 1,
          "rating": 1,
          "title": "title-1",
          "url": "url1",
        },
        Object {
          "description": "description-2",
          "id": 2,
          "price": 2,
          "rating": 2,
          "title": "title-2",
          "url": "url2",
        },
        Object {
          "description": "description-3",
          "id": 3,
          "price": 3,
          "rating": 3,
          "title": "title-3",
          "url": "url3",
        },
        Object {
          "description": "description-4",
          "id": 4,
          "price": 4,
          "rating": 4,
          "title": "title-4",
          "url": "url4",
        },
        Object {
          "description": "description-5",
          "id": 5,
          "price": 5,
          "rating": 5,
          "title": "title-5",
          "url": "url5",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": "reduxForm",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "currentUser": Object {
          "firstName": "beebee",
          "lastName": "sanders",
          "username": "bsanders",
        },
        "error": null,
        "handleSubmit": [Function],
        "loggedIn": false,
        "pristine": true,
        "reset": false,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "submitSucceeded": null,
        "submitting": false,
        "tools": Array [
          Object {
            "description": "description-1",
            "id": 1,
            "price": 1,
            "rating": 1,
            "title": "title-1",
            "url": "url1",
          },
          Object {
            "description": "description-2",
            "id": 2,
            "price": 2,
            "rating": 2,
            "title": "title-2",
            "url": "url2",
          },
          Object {
            "description": "description-3",
            "id": 3,
            "price": 3,
            "rating": 3,
            "title": "title-3",
            "url": "url3",
          },
          Object {
            "description": "description-4",
            "id": 4,
            "price": 4,
            "rating": 4,
            "title": "title-4",
            "url": "url4",
          },
          Object {
            "description": "description-5",
            "id": 5,
            "price": 5,
            "rating": 5,
            "title": "title-5",
            "url": "url5",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": "reduxForm",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
